### ISSUE TITLE
Improve embedded FlutterFragment's back button navigation delegation

### ISSUE THREAD

**Issue by math1man** [flutter/flutter#67011]
Add-to-app Flutter can be used in two ways. It can be embedded in its own dedicated activity, or it can share an activity with other native code. The former strategy works pretty well from what I've seen, however, it lacks a lot of flexibility because it means you can't really put Flutter and native UI side by side. The latter solution on the other hand has some issues. This issue involve navigation. 

The FlutterFragment documents that the activity should call through to the fragment's `onBackPressed()` method. This has two problems:

1. If the FlutterFragment is nested inside a stack of fragments, this call through can be pretty tedious.
2. It's difficult for the native code to decide whether back behavior should be handled natively, or by flutter.

Both of these issues can be resolved using AndroidX's [OnBackPressedCallback](https://developer.android.com/reference/androidx/activity/OnBackPressedCallback) API. They high level summary of the API is that it allows fragments to register an `OnBackPressedCallback` to the fragment's activity. This instantly solves problem 1, but problem 2 requires a bit more work. 

`OnBackPressedCallbacks` have an internal enabled state that controls whether the activity should delegate to the callback or to its own logic. (The callbacks are also associated with the fragment's lifecycle to ensure that they're automatically disabled when the fragment is detached, and there's also a complex algorithm to determine callback precedence if multiple are active simultaneously). It should be possible to add functionality to Flutter's navigation channel such that Dart code can call through to update the enabled state of the callback, resolving problem 2. There's a small chance for race conditions updating the enabled state because of the async nature of channels, but this is still a strict improvement. There are probably ways to improve the raciness too, I just haven't thought too much about them.

**Comment by xster** [comment]
To re-state the problem and make sure I understand, you're suggesting that we change Flutter's back button callback from a edge triggered signal (but also pre-filtered/piped by its native owner) to a level triggered signal (by having Flutter pre-report whether it has an inner stack). 

Did I capture it correctly?

**Comment by math1man** [comment]
Yes, I believe that's correct. The native code ought to know whether Flutter can handle back navigation in order for native code to decide whether to handle back navigation natively or delegate it to Flutter.

**Comment by hamdikahloun** [comment]
Hi @math1man 

Can you try this [project](https://github.com/hamdikahloun/FlutterFragment) and confirm if this is the right behavior .

> This will create a problem if a FlutterFragment subclass uses the OnBackPressedDispatcher to receive onBackPressed callbacks, which is a logical alternative to explicitly calling through from the activity (my app has already done this). I think that this change needs to account for that, and the best way is to have FlutterFragment's onBackPressed method explicitly triggered by an OnBackPressedCallback so that the Flutter framework can disable the callback in this specific case.

Thank you

**Comment by math1man** [comment]
No, that's not the correct behavior. You're still calling through from `MyActivity#onBackPressed` to `MyFlutterFragment#onBackPressed`. I'm specifically not doing that, and instead relying on the `OnBackPressedDispatcher to call `MyFlutterFragment#onBackPressed`.

**Comment by math1man** [comment]
A solution to this issue likely requires change to flutter android embedding code, because the Dart navigator needs to proactively (rather than reactively) inform the Android code whether it should handle back navigation. In other words, every forward or backward navigation event in Dart needs to update Android native state to appropriately enable/disable the `OnBackPressedCallback`.

**Comment by hamdikahloun** [comment]
Hi @math1man 

https://github.com/flutter/engine/pull/22495#discussion_r524464691

Can you please provide a minimal complete reproducible code sample ?

Thank you

**Comment by math1man** [comment]
Sure, see https://github.com/math1man/FlutterFragment/tree/master/android/app/src/main/java/com/example/fragmentbug

**Comment by hamdikahloun** [comment]
Can you please provide a minimal complete reproducible code sample for your issue ?

Thank you

**Comment by math1man** [comment]
Sorry, I misunderstood. I've updated https://github.com/math1man/FlutterFragment as a proper repro case.

**Comment by xster** [comment]
@math1man @hamdikahloun I'm gonna comment here since this will likely be the longest living bug. Since we might need to do a smaller scoped fix in https://github.com/flutter/engine/pull/22495 (and thus https://github.com/flutter/flutter/issues/69879 will be closed too). I think I understand the problem now after @math1man explained it to me. And https://github.com/flutter/engine/pull/22552 should indeed also be checked in with the fix to https://github.com/flutter/flutter/issues/69879 to make sure the interplay between the 2 bugs don't get messed up.

To describe the potential future "ideal state":

1. Flutter the SDK always reports back to the platform how deep its own navigation stack is. Possibly updating every time there's a change. Possibly further simplifiable into a binary 1 or >1. 
2. Through some platform channel / platform plugin / FlutterFragment plumbing, there's a default implementation in FlutterFragment that listens to this signal.
3. FlutterFragment takes this signal, and either adds its OnBackPressedCallback (which plumbs into platform channel, then Flutter) when Flutter's stack is >1, or removes its OnBackPressedCallback from the activity's OnBackPressedDispatcher. If it doesn't have a OnBackPressedCallback when Flutter's stack is 1, the activity will just remove the FlutterFragment and destroy the Flutter content by default.
4. Independently from 1-3, at any time Flutter can invoke SystemNavigator.pop programmatically without going through anything related to the back button. 
5. When it does, blindly dismissing the FlutterFragment's activity is incorrect in case there are more fragments in the activity (this bug). In that case, passing it to the OnBackPressedDispatcher's onBackPressed is appropriate. 
6a. To prevent that call from looping back to the FlutterFragment itself as possible via 3, the FlutterFragment should remove its own OnBackPressedCallback from the OnBackPressedDispatcher for the duration of this call. 
6b. The semantic of SystemNavigator.pop popping anything else besides Flutter itself might be weird...? Maybe it should just call fragmentManager.beginTransaction().remove(this).commit() instead? @math1man.

Since 1 is non-trivial (involves a lot of plumbing from flutter to engine), we might have to solve 1-3 independently from 4-6 without breaking 1-3. I think 6b might solve this bug without affecting 1-3? If we do 6a, perhaps a general refactor that's sensible to do either way is to not let all the control be in the PlatformPlugin itself. Instead, let it delegate back to an injected callback which the FlutterActivity and FlutterFragment implement. In the default implementations, the activity pops and the fragment does 6a. In case users are doing custom things with OnBackPressedDispatcher, they can subclass FlutterFragment, figure out when a pop is not from a back button, to disable the custom thing and prevent infinite loops. This way, it's backward compatible and configurable. @math1man does this make sense?

**Comment by math1man** [comment]
Thanks for the super thorough writeup. It covers most of the points accurately. I do want to mention that `OnBackPressedCallbacks` should be enabled and disabled rather than added and removed, but that's functionally equivalent.

As for 6b, I'd be very hesitant to do that. The main reason is that it may not respect what the native code actually wants to do. Maybe the native code *does* want to finish the activity. Or maybe the native code has some other navigation entirely (consider a tab architecture where one tab is flutter, and back navigation should return to a different tab).

Perhaps a viable short-term solution to set the groundwork for 1-3 would be to create some sort of method or interface like `getOnBackPressedCallback` in the FlutterFragment or somewhere related. This way, the method can be optionally implemented (such as by my app) and @hamdikahloun 's PR can access it to do the correct thing (if necessary) without plumbing everything through the dart layers. Later, the resolution of 1-3 can provide a definitive implementation of this method.

In this way, the resolution to 4-6, albeit a breaking change from my perspective, can be very easily be resolved by implementing the new method. Later, I can clean up the implementation.

**Comment by xster** [comment]
> I do want to mention that OnBackPressedCallbacks should be enabled and disabled rather than added and removed, but that's functionally equivalent.

oops

> As for 6b, I'd be very hesitant to do that.

Whether we do 6a or 6b, the point I was trying to make in the next paragraph is we should plumb the callback from the PlatformPlugin to the FlutterFragment or activity either way. Therefore 6a or 6b is just a matter of a default behavior that's overridable by the subclass. 

> Perhaps a viable short-term solution to set the groundwork for 1-3 would be to create some sort of method or interface like getOnBackPressedCallback in the FlutterFragment or somewhere related. This way, the method can be optionally implemented (such as by my app) and @hamdikahloun 's PR can access it to do the correct thing (if necessary) without plumbing everything through the dart layers.

I think we're saying the same thing? Though what I'm saying is to let that callback be called from FlutterFragment on the way back from Flutter via popSystemNavigator. You still do whatever work-around you're currently doing for 1-3. By overriding this in 4-6 in your app, you can decide to let it do something or nothing depending on whether it was a back button press or just an independent SystemNavigator.pop call from dart.

**Comment by math1man** [comment]
I must have misread 6a, because I think that's exactly what I think needs to happen. Maybe a simple solution is to create some interface like `SystemNavigationPoppable` or something that both FlutterFragment and FlutterActivity can implement, and pass that to PlatformPlugin constructor so that the fragment or activity can define the behavior rather than burying it inaccessibly inside of PlatformPlugin. That way we can implement logical defaults for each class, but I can still override it, and a future change can also leverage that for implementing 1-3.

**Comment by xster** [comment]
> Maybe a simple solution is to create some interface like SystemNavigationPoppable or something that both FlutterFragment and FlutterActivity can implement, and pass that to PlatformPlugin constructor so that the fragment or activity can define the behavior rather than burying it inaccessibly inside of PlatformPlugin

Right. That's exactly what I meant as well. I'm not 100% on the name yet. Maybe we want to make it not so single-purposed (`FlutterPlatformDelegate` or some such).

**Comment by math1man** [comment]
SGTM. At first I thought maybe make it part of FlutterActivityAndFragmentDelegate.Host, but this needs to be public and you probably want to keep that private.

**Comment by xster** [comment]
Generally prefer sparse composition rather than big interfaces. `FlutterActivityAndFragmentDelegate.Host` might not help since we wouldn't want to leak that concept into the PlatformPlugin. 

@hamdikahloun are you still interested in taking this on? :D otherwise I can make that PR too.

**Comment by hamdikahloun** [comment]
@xster I think you are more familiar with that code than I , but I will follow this issue to improve my knowledge .

**Comment by math1man** [comment]
I'm wondering if this issue can be resolved following https://github.com/flutter/engine/pull/25453

**Comment by xster** [comment]
Thanks for fixing it Ari!

**Comment by xster** [comment]
Actually I shouldn't close it. We should still aim for the ideal solution (solving 1-3 in https://github.com/flutter/flutter/issues/67011#issuecomment-730665542)

**Comment by xster** [comment]
For the remaining 1-3, might make sense to add to https://cs.opensource.google/flutter/engine/+/master:shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java;l=56;drc=a887ae9e27ae1a992b29f17a8c722151a74830ed and pre-emptively buffer whether the current Flutter stack is 1 screen deep or deeper. Though whether there could be a race going from 3->2->1 (pop Flutter so far)->pop native should be tested. Otherwise, we might need to make sure the buffer on the platform is thread safe, but then we need to solve https://github.com/flutter/flutter/issues/29081 first

**Comment by wahyupermadie** [comment]
Hi team, I have a similar problem with this [issue](https://github.com/flutter/flutter/issues/115768), Are we already able to resolve this?
